saleor/graphql/core/connection.py

```diff
@@ -1,7 +1,168 @@
+    field_path = field.split("__")
+    attr = instance
+    for elem in field_path:
+        try:
+            attr = getattr(attr, elem)
+        except AttributeError:
+            return None
+    return attr
+
+
+def get_field_value(instance, field_name):
+    return get_repr(get_field(instance, field_name))
+
+
+def connection_args_validation(args):
+    first = args.get("first")
+    last = args.get("last")
+
+    if isinstance(first, int) and first < 0 or first and not isinstance(first, int):
+        raise ValueError("Argument 'first' must be a non-negative integer.")
+    if isinstance(last, int) and last < 0 or last and not isinstance(last, int):
+        raise ValueError("Argument 'last' must be a non-negative integer.")
+    if first and last:
+        raise ValueError("Argument 'last' cannot be combined with 'first'.")
+    if first and args.get("before"):
+        raise ValueError("Argument 'first' cannot be combined with 'before'.")
+    if last and args.get("after"):
+        raise ValueError("Argument 'last' cannot be combined with 'after'.")
+
+
+def connection_from_queryset_slice(
+    qs: QuerySet,
+    args: ConnectionArguments = None,
+    connection_type: typing.Any = Connection,
+    edge_type: typing.Any = Edge,
+    page_info_type: typing.Any = PageInfo,
+) -> Connection:
+    """Create a connection object from a QuerySet."""
+    args = args or {}
+    before = args.get("before")
+    after = args.get("after")
+    first = args.get("first")
+    last = args.get("last")
+    connection_args_validation(args)
+
+    cursor_after = from_global_cursor(after) if after else None
+    cursor_before = from_global_cursor(before) if before else None
+    cursor = cursor_after or cursor_before
+    cursor_offset = 1 if cursor else 0
+
+    sort_by = args.get("sort_by", {})
+    sorting_direction = sort_by.get("direction", "")
+    sorting_fields = sort_by.get("field")
+    if sorting_fields and not isinstance(sorting_fields, list):
+        sorting_fields = [sorting_fields]
+    elif not sorting_fields:
+        sorting_fields = []
+
+    requested_count = first or last
+    end_margin = requested_count + cursor_offset + 1 if requested_count else None
+
+    if last:
+        # reversed direction
+        sorting_direction_kw = (
+            "gte" if sorting_direction == OrderDirection.DESC else "lte"
+        )
+    else:
+        sorting_direction_kw = (
+            "lte" if sorting_direction == OrderDirection.DESC else "gte"
+        )
+
+    filter_kwargs = Q()
+    if cursor and sorting_fields:
+        for field_id in range(len(sorting_fields)):
+            field_expression = {}
+            for val_id in range(len(cursor[:field_id])):
+                field_expression[sorting_fields[val_id]] = cursor[val_id]
+            field_expression[
+                f"{sorting_fields[field_id]}__{sorting_direction_kw}"
+            ] = cursor[field_id]
+            filter_kwargs |= Q(**field_expression)
```

Comment:
Not sure this is correct, for sorting fields `["foo", "bar", "baz"]` and cursor values `[1, 2, 3]` we should filter by:

```python
Q(foo__gt=1) | Q(foo=1, bar__gt=2) | Q(foo=1, bar=2, baz__gt=3)
```

Reply:
that's what i'm doing
first have for loop for every index in sorting_fields list
then we take index we currently have and use it to slice sorting_fields (i.e. index 1 will slice sorting_fields so it will be [], index 2 will slice sorting_fields so it will be [field_1] etc.) and we run through this list and create `Q(foo=1),  Q(foo=1, bar=2)` etc
after that we append comparing condition with last field from loop
and we repeat this for every field in sorting_fields list

I know it's hard to read, that's why I will try to simplify this, or move this to separate method

Reply:
I think some comment blocks would also help a lot 😄

Reply:
Agree, they would help a lot.